dll/win32/comctl32/static.c

```diff
@@ -627,6 +627,18 @@ static void STATIC_PaintTextfn( HWND hwnd, HDC hdc, DWORD style )
     else
     {
         DrawTextW( hdc, text, -1, &rc, format );
+        if(style & WS_DISABLED)
+        {
+            COLORREF oldTextColor=GetTextColor(hdc);
+            SetBkMode(hdc, TRANSPARENT);
+            SetTextColor(hdc, GetSysColor(COLOR_HIGHLIGHTTEXT));
+            rc.left++;rc.top++;
+            DrawTextW( hdc, text, -1, &rc, format );
```

Comment:
Isn't the DrawTextW call that is before `if(style & WS_DISABLED)` redundant in this case?

Reply:
If it is omitted, the background of the text is not drawn. It would have to be replaced by rendering a rectangle, but I don't know if it's easy to render a rectangle the width of the text, (with NSIS installer it doesn't matter, there the background is already rendered in the dialog, but somewhere it can be used with a background).

Reply:
And if we render the background while drawing the shadow, it will be 1px offset.

Reply:
Using the same way as done by comctl32/user32 button.c (see my links above) should solve the problem.

Reply:
I'll take another look at it, but maybe in a week, I'm on vacation now.

1)The specification says DSS_DISABLED, but the NSIS installer has WS_DISABLED in *.rc. So it would only work if they had the same value, but we have WS_DISABLED 0x8000000 and DSS_DISABLED 32.

2)
In button.c is just this:
if (style & WS_DISABLED) flags |= DFCS_INACTIVE;

The rendering is in draw.c - I'll try to see if I can use that instead of DrawTextW.
MyIntDrawRectEdge(dc, r, (uFlags & DFCS_PUSHED) ? EDGE_SUNKEN : EDGE_RAISED, (uFlags&DFCS_FLAT) | BF_MIDDLE | BF_RECT, scheme);
    ZeroMemory(&lf, sizeof(lf));
    MyMakeSquareRect(r, &myr);
    myr.left += 1;
    myr.top += 1;
    myr.right -= 1;
    myr.bottom -= 1;
    if (uFlags & DFCS_PUSHED)
       OffsetRect(&myr,1,1);
    lf.lfHeight = myr.bottom - myr.top;
    lf.lfWidth = 0;
    lf.lfWeight = FW_NORMAL;
    lf.lfCharSet = DEFAULT_CHARSET;
    lstrcpy(lf.lfFaceName, TEXT("Marlett"));
    hFont = CreateFontIndirect(&lf);
    /* Save font and text color */
    hOldFont = SelectObject(dc, hFont);
    clrsave = GetTextColor(dc);
    bkmode = GetBkMode(dc);
    /* Set color and drawing mode */
    SetBkMode(dc, TRANSPARENT);
    if (uFlags & DFCS_INACTIVE)
    {
        /* Draw shadow */
        SetTextColor(dc, scheme->crColor[COLOR_BTNHIGHLIGHT]);
        TextOut(dc, myr.left + 1, myr.top + 1, &Symbol, 1);
    }
    SetTextColor(dc, scheme->crColor[(uFlags & DFCS_INACTIVE) ? COLOR_BTNSHADOW : COLOR_BTNTEXT]);
    /* Draw selected symbol */
    TextOut(dc, myr.left, myr.top, &Symbol, 1);
    /* restore previous settings */
    SetTextColor(dc, clrsave);
    SelectObject(dc, hOldFont);
    SetBkMode(dc, bkmode);
    DeleteObject(hFont);

Reply:
This looks much more complicated than what I was mentioning. Namely, calling DrawStateW for strings. (and we know that buttons can correctly draw that grayed text, and they use DrawStateW)

Reply:
There is one more place that uses DSS_DISABLE(also in draw.c), but it's not much easier(otherwise I haven't found anything, the button rendering is done in one of these places). The rendering is in draw.c - I'll try to see if I can use that instead of DrawTextW. Anyway, I'll try to run both and figure something out - use BitBlt, etc.

DrawStateW->IntDrawState
    if (flags & (DSS_MONO | DSS_DISABLED))
    {
        /* Create a non-antialiased font */
        GetObjectW(hFontOriginal, sizeof(lf), &lf);
        lf.lfQuality = NONANTIALIASED_QUALITY;
        hNaaFont = CreateFontIndirectW(&lf);
    }
.........
    if (flags & DSS_DISABLED)
        hbrtmp = GetSysColorBrush(COLOR_3DHILIGHT);
    else if (flags & DSS_DEFAULT)
        hbrtmp = GetSysColorBrush(COLOR_3DSHADOW);

    /* Draw light or dark shadow */
    if (flags & (DSS_DISABLED|DSS_DEFAULT))
    {
        if(!hbrtmp) goto cleanup;
        hbsave = (HBRUSH)SelectObject(hdc, hbrtmp);
        if(!hbsave) goto cleanup;
        if(!BitBlt(hdc, x+1, y+1, cx, cy, memdc, 0, 0, 0x00B8074A)) goto cleanup;
        SelectObject(hdc, hbsave);
    }

    if (flags & DSS_DISABLED)
    {
        hbr = hbrtmp = GetSysColorBrush(COLOR_3DSHADOW);
        if(!hbrtmp) goto cleanup;
    }
    else if (!hbr)
    {
        hbr = (HBRUSH)GetStockObject(BLACK_BRUSH);
    }

Reply:
Too much complicated. As I implicitly evoked there are already existing functions that do the job correctly, and even mentioned that they were used in the button.c code I linked below.